File: item-list/src/main/java/com/gurunars/item_list/ItemList.java
package com.gurunars.item_list;

import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import java.util.List;

import butterknife.ButterKnife;

/**
 * A wrapper around RecyclerView with reasonable default item change animations and a
 * composition based solution as a mean to bind data to item list views.
 *
 * @param <PayloadType> Item payload type
 */
public class ItemList<PayloadType extends Payload> extends FrameLayout {

    private ItemAdapter<PayloadType> itemAdapter;
    private LinearLayoutManager layoutManager;

    public ItemList(Context context) {
        this(context, null);
    }

    public ItemList(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ItemList(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        inflate(context, R.layout.item_list, this);

        RecyclerView recyclerView = ButterKnife.findById(this, R.id.recyclerView);
        layoutManager = new LinearLayoutManager(context);
        itemAdapter = new ItemAdapter<>(new Scroller() {
            @Override
            public void scrollToPosition(int position) {
                layoutManager.scrollToPosition(position);
            }

            @Override
            public int findFirstVisibleItemPosition() {
                return layoutManager.findFirstVisibleItemPosition();
            }

            @Override
            public int findLastVisibleItemPosition() {
                return layoutManager.findLastVisibleItemPosition();
            }
        });

        recyclerView.setAdapter(itemAdapter);
        recyclerView.setLayoutManager(layoutManager);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);

    }

    /**
     * Set a collection of items to be shown.
     *
     * If the collection of items is different - the diff shall be animated.
     *
     * @param items a new collection to be shown
     */
    public void setItems(List<Item<PayloadType>> items) {
        itemAdapter.setItems(items);
    }

    /**
     * Map item type to view binder responsible for rending items of this type.
     *
     * @param itemType type of the Item
     * @param itemViewBinder renderer for the items of a given type
     */
    public void registerItemViewBinder(Enum itemType, ItemViewBinder<PayloadType> itemViewBinder) {
        itemAdapter.registerItemViewBinder(itemType, itemViewBinder);
    }

    /**
     * Set the renderer to be employed when the list contains no items.
     *
     * @param emptyViewBinder renderer for the empty list
     */
    public void setEmptyViewBinder(EmptyViewBinder emptyViewBinder) {
        itemAdapter.setEmptyViewBinder(emptyViewBinder);
    }

}

File: item-list/src/main/kotlin/com/gurunars/item_list/ItemList.kt
package com.gurunars.item_list

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import com.esotericsoftware.kryo.Kryo
import com.gurunars.databinding.BindableField
import com.gurunars.databinding.bindableField
import com.gurunars.shortcuts.fullSize
import com.gurunars.shortcuts.setPadding
import org.jetbrains.anko.bottomPadding
import org.jetbrains.anko.dip
import org.jetbrains.anko.recyclerview.v7.recyclerView
import org.objenesis.strategy.StdInstantiatorStrategy


internal fun defaultEmptyViewBinder(context: Context): View {
    return TextView(context).apply {
        fullSize()
        setText(R.string.empty)
        gravity = Gravity.CENTER
    }
}


private fun<ItemType> defaultItemViewBinder(context: Context, payload: BindableField<Pair<ItemType?, ItemType?>>) : View {
    return TextView(context).apply {
        setPadding(context.dip(5))
        payload.onChange {
            text = it.first.toString()
        }
    }
}


class ItemList<ItemType : Item> constructor(context: Context) : FrameLayout(context) {

    private val kryo = Kryo().apply {
        instantiatorStrategy = Kryo.DefaultInstantiatorStrategy(StdInstantiatorStrategy())
    }

    val items = bindableField(
        listOf<ItemType>(),
        {one, two -> equal(one, two) },
        {item -> kryo.copy(ArrayList(item))}
    )

    val emptyViewBinder = bindableField<EmptyViewBinder>(::defaultEmptyViewBinder)
    val itemViewBinders = bindableField<Map<Enum<*>, ItemViewBinder<ItemType>>>(mapOf())
    val defaultViewBinder = bindableField<ItemViewBinder<ItemType>>(::defaultItemViewBinder)

    init {

        recyclerView {
            id=R.id.recyclerView
            fullSize()
            clipToPadding=false
            bottomPadding=dip(60)
            isSaveEnabled=false
            adapter = ItemAdapter(
                items,
                emptyViewBinder,
                itemViewBinders,
                defaultViewBinder
            )
            layoutManager = LinearLayoutManager(context).apply {
                orientation = LinearLayoutManager.VERTICAL
            }
        }

    }

}
